pkg/kubelet/pod/mirror_manager.go

```diff
@@ -0,0 +1,373 @@
+	if pod, ok := mp.pods[key]; ok {
+		delete(mp.upods, pod.UID)
+		delete(mp.pods, key)
+	}
+}
+
+type mirrorManager struct {
+	// Protects all the data in mirrorManager
+	lock sync.RWMutex
+	// Static pods indexed by full name
+	staticPods map[string]*api.Pod
+	// Mirror pods from api server
+	mirrorPods mirrorPods
+
+	// The channel receiving mirror pod to be created
+	mirrorPodCreationChannel chan *api.Pod
+	// The channel receiving mirror pod to be deleted
+	mirrorPodDeletionChannel chan *api.Pod
+
+	// apiserver client
+	apiserverClient clientset.Interface
+}
+
+const syncPeriod = 5 * time.Second
+
+func newMirrorManager(apiserverClient clientset.Interface) *mirrorManager {
+	return &mirrorManager{
+		staticPods:      make(map[string]*api.Pod),
+		mirrorPods:      newMirrorPods(),
+		apiserverClient: apiserverClient,
+		// There shouldn't be too many mirror pod changes, set buffer size to 100
+		mirrorPodCreationChannel: make(chan *api.Pod, 100),
+		mirrorPodDeletionChannel: make(chan *api.Pod, 100),
+	}
+}
+
+func (m *mirrorManager) Start() {
+	// Don't start mirror pod manager goroutine if we don't have a client. This will
+	// happen on the master, where the kubelet is responsible for bootstrapping the
+	// pods of the master components.
+	if m.apiserverClient == nil {
+		glog.Infof("Apiserver client is nil, not starting mirror pod manager.")
+		return
+	}
+
+	glog.Info("Starting mirror pod manager syncing loop")
+	syncTicker := time.Tick(syncPeriod)
+	go wait.Forever(func() {
+		select {
+		case pod := <-m.mirrorPodCreationChannel:
+			m.handleMirrorPodCreation(pod)
+		case pod := <-m.mirrorPodDeletionChannel:
+			m.handleMirrorPodDeletion(pod)
+		case <-syncTicker:
+			m.handleCleanup()
+		}
+	}, 0)
+}
+
+// Add and Delete a static pod, and the corresponding mirror pod will be created, recreated
+// and deleted. These two functions should be only used in pod manager now.
+// We'll always create mirror pod when addStaticPod(), the caller should make sure that the static pod
+// is really changed. Usually kubelet config will guarantee that.
+func (m *mirrorManager) addStaticPod(staticPod *api.Pod) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.staticPods[getKey(staticPod)] = staticPod
+	glog.V(2).Infof("Static pod %q is added or updated, create new mirror pod", format.Pod(staticPod))
+	m.createMirrorPod(staticPod)
+}
+
+func (m *mirrorManager) deleteStaticPod(staticPod *api.Pod) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	key := getKey(staticPod)
+	if _, found := m.staticPods[key]; !found {
+		glog.V(4).Infof("The static pod %q is not found", format.Pod(staticPod))
+		return
+	}
+	glog.V(4).Infof("Delete static pod %q, delete the mirror pod", format.Pod(staticPod))
+	delete(m.staticPods, key)
+	// In fact using static pod directly is also ok, but it's more intuitive making sure that
+	// only mirror pod is passing into deleteMirrorPod() and finally handleMirrorPodDeletion()
+	m.deleteMirrorPod(newMirrorPod(staticPod))
```

Comment:
Can we check if there is an existing mirror pod that needs to be deleted? Calling `newMirrorPod` is a bit strange.